File: windows/RecordHarness/DeviceBoundClient.cs
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using PabloCompanion.Core;

namespace RecordHarness;

/// <summary>
/// Device-bound HTTP plumbing for the headless scenario.
///
/// Owns the enrollment handshake (<c>/api/auth/native/code</c> →
/// <c>/exchange</c> carrying a real enrollment payload) and DPoP-signed requests
/// (<c>Bearer</c> + <c>X-Install-ID</c> + a fresh proof), so the scenario drives
/// the identical device-binding code the shipping app uses. Every signed request
/// produces its proof from the real <see cref="DpopProof"/>, so a drift between
/// the client crypto and the server's DPoP middleware fails here.
///
/// The key is an ephemeral in-memory P-256 keypair with a fresh install id per
/// run: no credential vault, nothing persisted, and no collision with a real
/// install on the same machine.
///
/// The C# port of the harness's Swift <c>DeviceBoundClient</c>.
/// </summary>
public sealed class DeviceBoundClient : IDisposable
{
    /// <summary>
    /// The redirect URI only has to be an allowed native scheme — the harness
    /// never opens it; the code comes back in the POST body.
    /// </summary>
    public const string DefaultRedirectUri = "pablohealth://auth/callback";

    private readonly string _baseUrl;
    private readonly HttpClient _http;
    private readonly ECDsa _deviceKey;

    public string InstallId { get; }

    public DeviceBoundClient(string baseUrl, string installId, HttpClient? http = null)
    {
        _baseUrl = baseUrl.TrimEnd('/');
        InstallId = installId;
        _http = http ?? new HttpClient();
        _deviceKey = ECDsa.Create(ECCurve.NamedCurves.nistP256);
    }

    public void Dispose() => _deviceKey.Dispose();

    /// <param name="Status">HTTP status code.</param>
    /// <param name="Body">Raw response body.</param>
    public sealed record Response(int Status, string Body)
    {
        /// <summary>The body parsed as JSON, or null when it isn't JSON.</summary>
        public JsonElement? Json
        {
            get
            {
                try
                {
                    using var doc = JsonDocument.Parse(Body);
                    return doc.RootElement.Clone();
                }
                catch (JsonException)
                {
                    return null;
                }
            }
        }

        public string BodyPrefix => Body.Length <= 200 ? Body : Body[..200];
    }

    /// <param name="IdToken">Refreshed id token returned by /exchange (falls back to the input).</param>
    /// <param name="ExchangeStatus">HTTP status of the /exchange call (200 on success).</param>
    /// <param name="KeyStorage">key_storage reported in the enrollment payload.</param>
    public sealed record Enrollment(string IdToken, int ExchangeStatus, string KeyStorage);

    // --- Enrollment ---

    /// <summary>
    /// Runs the OAuth code exchange carrying a real device-enrollment payload,
    /// returning the refreshed id token. Throws if <c>/native/code</c> fails.
    /// </summary>
    public async Task<Enrollment> EnrollAsync(
        string idToken,
        string refreshToken,
        string redirectUri = DefaultRedirectUri,
        CancellationToken cancellationToken = default)
    {
        var code = await PostJsonAsync("/api/auth/native/code", new
        {
            id_token = idToken,
            refresh_token = refreshToken,
            redirect_uri = redirectUri,
        }, cancellationToken);

        var oneTimeCode = code.Json.GetStringOrNull("code");
        if (code.Status != 200 || string.IsNullOrEmpty(oneTimeCode))
            throw new HarnessException($"native/code failed: {code.Status} {code.BodyPrefix}");

        // The real enrollment payload, built by the shared core from this run's
        // ephemeral key — the same builder the app uses with its vault key.
        var enrollment = DeviceEnrollment.BuildPayload(
            InstallId,
            DeviceEnrollment.PublicJwk(_deviceKey),
            DeviceEnrollment.SoftwareKeyStorage);

        var exchange = await PostJsonAsync("/api/auth/native/exchange", new
        {
            code = oneTimeCode,
            redirect_uri = redirectUri,
            enrollment,
        }, cancellationToken);

        var refreshed = exchange.Json.GetStringOrNull("id_token") ?? idToken;
        return new Enrollment(refreshed, exchange.Status, DeviceEnrollment.SoftwareKeyStorage);
    }

    // --- HTTP ---

    /// <summary>Unauthenticated JSON POST (the native code/exchange endpoints).</summary>
    public async Task<Response> PostJsonAsync(
        string path, object body, CancellationToken cancellationToken = default)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, _baseUrl + path)
        {
            Content = new StringContent(JsonSerializer.Serialize(body), Encoding.UTF8, "application/json"),
        };
        return await SendAsync(request, cancellationToken);
    }

    /// <summary>
    /// Authenticated request carrying the device binding exactly like the app:
    /// <c>Bearer</c> + <c>X-Install-ID</c> + a fresh <c>DPoP</c> proof.
    /// </summary>
    public async Task<Response> RequestAsync(
        HttpMethod method,
        string path,
        string idToken,
        object? jsonBody = null,
        CancellationToken cancellationToken = default)
    {
        var url = _baseUrl + path;
        using var request = new HttpRequestMessage(method, url);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", idToken);
        request.Headers.Add("X-Install-ID", InstallId);
        request.Headers.Add("DPoP", DpopProof.Create(_deviceKey, method.Method, url));

        if (jsonBody is not null)
        {
            request.Content = new StringContent(
                JsonSerializer.Serialize(jsonBody), Encoding.UTF8, "application/json");
        }

        return await SendAsync(request, cancellationToken);
    }

    /// <summary>
    /// Stamps the device binding onto a request built elsewhere — the delegate the
    /// shared <see cref="AudioUploadClient"/> takes, so the harness's upload carries
    /// the same headers as the app's.
    /// </summary>
    public void AttachBinding(HttpRequestMessage request)
    {
        if (request.RequestUri is null) return;
        request.Headers.Add("DPoP", DpopProof.Create(
            _deviceKey, request.Method.Method, request.RequestUri.ToString()));
        request.Headers.Add("X-Install-ID", InstallId);
    }

    private async Task<Response> SendAsync(
        HttpRequestMessage request, CancellationToken cancellationToken)
    {
        using var response = await _http.SendAsync(request, cancellationToken);
        var body = await response.Content.ReadAsStringAsync(cancellationToken);
        return new Response((int)response.StatusCode, body);
    }

    /// <summary>ISO-8601 with an internet date-time format, matching the Swift harness.</summary>
    public static string Iso8601(DateTimeOffset date) =>
        date.ToUniversalTime().ToString("yyyy-MM-ddTHH:mm:ssZ");
}

File: Sources/practice-harness/DeviceBoundClient.swift
#if canImport(CompanionAuthCore)

import CompanionAuthCore
import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// Shared device-bound HTTP plumbing for the headless scenarios.
///
/// Owns the enrollment handshake (`/api/auth/native/code` → `/exchange` carrying
/// a real `DeviceEnrollment` payload) and DPoP-signed requests (`Bearer` +
/// `X-Install-ID` + a fresh `DPoPProof`), so the `dpop` and `record` scenarios
/// drive the identical device-binding code the shipping app uses. Every signed
/// request produces its proof from the actual `DPoPProof.make`, so a drift
/// between the client crypto and `backend/app/middleware/dpop.py` fails here.
struct DeviceBoundClient {
    let baseURL: String
    let installID: String

    struct Response {
        let status: Int
        let body: Data
        var json: [String: Any]? {
            (try? JSONSerialization.jsonObject(with: body)) as? [String: Any]
        }

        var bodyPrefix: String {
            String(decoding: body.prefix(200), as: UTF8.self)
        }
    }

    /// Result of the enrollment handshake.
    struct Enrollment {
        /// Refreshed id token returned by `/exchange` (falls back to the input).
        let idToken: String
        /// HTTP status of the `/exchange` call (200 on success).
        let exchangeStatus: Int
        /// `key_storage` reported by the device key (`secureEnclave` / `software`).
        let keyStorage: String
    }

    /// The redirect URI only has to be an allowed native scheme — the harness
    /// never opens it; the code comes back in the POST body.
    static let defaultRedirectURI = "pablohealth://auth/callback"

    // MARK: - Enrollment

    /// Runs the OAuth code exchange carrying a real device-enrollment payload,
    /// returning the refreshed id token. Throws if `/native/code` fails or no
    /// device key can be provisioned.
    func enroll(
        idToken: String,
        refreshToken: String,
        redirectURI: String = defaultRedirectURI
    ) async throws -> Enrollment {
        let code = try await postJSON(
            path: "/api/auth/native/code",
            body: ["id_token": idToken, "refresh_token": refreshToken, "redirect_uri": redirectURI]
        )
        guard let oneTimeCode = code.json?["code"] as? String, code.status == 200 else {
            throw DeviceBoundError("native/code failed: \(code.status) \(code.bodyPrefix)")
        }

        guard let enrollment = DeviceEnrollment.payload(installID: installID) else {
            throw DeviceBoundError("DeviceEnrollment.payload returned nil — no device key could be provisioned")
        }
        let keyStorage = (enrollment["key_storage"] as? String) ?? "?"

        let exchange = try await postJSON(
            path: "/api/auth/native/exchange",
            body: ["code": oneTimeCode, "redirect_uri": redirectURI, "enrollment": enrollment]
        )
        let refreshed = (exchange.json?["id_token"] as? String) ?? idToken
        return Enrollment(idToken: refreshed, exchangeStatus: exchange.status, keyStorage: keyStorage)
    }

    // MARK: - HTTP

    /// Unauthenticated JSON POST (the native code/exchange endpoints).
    func postJSON(path: String, body: [String: Any]) async throws -> Response {
        guard let url = URL(string: baseURL + path) else { throw DeviceBoundError("bad url \(path)") }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        return try await send(request)
    }

    /// Authenticated request carrying the device binding exactly like the app:
    /// `Bearer` + `X-Install-ID` + a fresh `DPoP` proof (or a caller-supplied
    /// proof for replay tests, or no proof at all for the negative test).
    func request(
        _ method: String,
        path: String,
        idToken: String,
        jsonBody: [String: Any]? = nil,
        presetProof: String? = nil,
        omitProof: Bool = false
    ) async throws -> Response {
        guard let url = URL(string: baseURL + path) else { throw DeviceBoundError("bad url \(path)") }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("Bearer \(idToken)", forHTTPHeaderField: "Authorization")
        request.setValue(installID, forHTTPHeaderField: "X-Install-ID")
        if !omitProof {
            guard let proof = presetProof ?? DPoPProof.make(method: method, url: url) else {
                throw DeviceBoundError("DPoPProof.make returned nil for \(method) \(path)")
            }
            request.setValue(proof, forHTTPHeaderField: "DPoP")
        }
        if let jsonBody {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: jsonBody)
        }
        return try await send(request)
    }

    private func send(_ request: URLRequest) async throws -> Response {
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw DeviceBoundError("non-HTTP response from \(request.url?.path ?? "?")")
        }
        return Response(status: http.statusCode, body: data)
    }

    // MARK: - Helpers

    func iso8601(_ date: Date) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.string(from: date)
    }
}

struct DeviceBoundError: LocalizedError {
    let message: String
    init(_ message: String) {
        self.message = message
    }

    var errorDescription: String? {
        message
    }
}

#endif
